bser.c

```diff
@@ -0,0 +1,163 @@
+#include <string.h>
+#include <stdarg.h>
+#include <assert.h>
+
+#include "bser.h"
+#include "bser_private.h"
+
+bser_t* bser_new_integer(uint64_t value, bser_t* fill) {
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_INT64;
+    fill->value.integer = value;
+    return fill;
+}
+
+bser_t* bser_new_real(double value, bser_t* fill) {
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_REAL;
+    fill->value.real = value;
+    return fill;
+}
+
+bser_t* bser_new_true(bser_t* fill) {
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_TRUE;
+    return fill;
+}
+
+bser_t* bser_new_false(bser_t* fill) {
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_FALSE;
+    return fill;
+}
+
+bser_t* bser_new_null(bser_t* fill) {
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_NULL;
+    return fill;
+}
+
+bser_t* bser_new_string(const char* chars, size_t len, bser_t* fill) {
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_STRING;
+    fill->value.string.chars = chars;
+    fill->value.string.length = len;
+    return fill;
+}
+
+bser_t* bser_new_array(
+        bser_t* elems, size_t len, bser_t* fill) {
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_ARRAY;
+    fill->value.array.elements = elems;
+    fill->value.array.length = len;
+    return fill;
+}
+
+bser_t* bser_new_object(
+        bser_key_value_pair_t* fields, size_t length, bser_t* fill) {
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_OBJECT;
+    fill->value.object.fields = fields;
+    fill->value.object.length = length;
+    return fill;
+}
+
+static void bser_free_array(bser_t* bser) {
+    for (int i = 0; i < bser->value.array.length; ++i) {
+        bser_free(&bser->value.array.elements[i]);
+    }
+}
+
+static void bser_free_object(bser_t* bser) {
+    for (int i = 0; i < bser->value.object.length; ++i) {
+        bser_free(&bser->value.object.fields[i].value);
+    }
+}
+
+void bser_free(bser_t* bser) {
+    if (bser_is_array(bser)) {
+        bser_free_array(bser);
+    } else if (bser_is_object(bser)) {
+        bser_free_object(bser);
+    }
+}
+
+char* bser_string_value_new(bser_t* bser) {
+    bser_parse_if_necessary(bser);
+    return strndup(bser->value.string.chars, bser->value.string.length);
+}
+
+int bser_string_strcmp(const char* match, bser_t* bser) {
+    bser_parse_if_necessary(bser);
+    int cmp = strncmp(match, bser->value.string.chars,
+                      bser->value.string.length);
+    return cmp == 0 ? match[bser->value.string.length] - '\0' : cmp;
```

Comment:
and actually, if chars are signed (which, on my machine, they are) and match[bser->value.string.length] < 0 (that is >127), this will return a negative number instead of a positive one.  So you probably need some casting.